python/hail/expr/expressions/expression_utils.py

```diff
@@ -226,3 +226,11 @@ def check_entry_indexed(caller, expr):
     if expr._indices != expr._indices.source._entry_indices:
         raise ExpressionException("{}: expression must be entry-indexed,"
                                   " found indices {}".format(caller, list(expr._indices.axes)))
+
+
+@typecheck(caller=str,
+           expr=Expression)
+def check_row_indexed(caller, expr):
+    if expr._indices != expr._indices.source._row_indices:
```

Comment:
Good point, I've replaced with `analyze('locus_windows', locus_expr, locus_expr._indices.source._row_indices)`, etc

Reply:
This was complicated because the expression may not have a source (which we want to catch with a good error message) and because analyze considers global expressions to be valid row expressions, which we do not want to be the case here. So instead I've fixed this function as well as `check_entry_indexed` to have the right behavior even when no source is present.